relay-redis/src/real.rs

```diff
@@ -382,3 +396,154 @@ pub struct Stats {
+    {
+        let client = self.client.clone();
+        Box::pin(async move { client.get_async_connection().await })
+    }
+
+    fn is_valid<'life0, 'life1, 'async_trait>(
+        &'life0 self,
+        conn: &'life1 mut Self::Connection,
+    ) -> Pin<Box<dyn Future<Output = Result<(), Self::Error>> + Send + 'async_trait>>
+    where
+        'life0: 'async_trait,
+        'life1: 'async_trait,
+    {
+        Box::pin(async move { redis::cmd("PING").query_async(conn).await })
+    }
+
+    fn has_broken(&self, _: &mut Self::Connection) -> bool {
+        false
+    }
+}
+
+/// An abstraction over Single and Cluster redis connections.
+#[derive(Debug, Clone)]
+pub enum AsyncRedisPool {
+    /// Represents a connection pool to a clustered redis instance.
+    Cluster(bb8::Pool<RedisClusterConnectionManager>, RedisConfigOptions),
+    /// Represents a connection pool to a single redis server.
+    Single(
+        bb8::Pool<bb8_redis::RedisConnectionManager>,
+        RedisConfigOptions,
+    ),
+}
+
+impl AsyncRedisPool {
+    /// Takes a command and executes it on redis.
+    pub async fn query_async<T: FromRedisValue>(&self, cmd: Cmd) -> Result<T, RedisError> {
+        match self {
+            Self::Cluster(pool, options) => {
+                let mut conn = pool.get().await.map_err(RedisError::AsyncPool)?;
+                timeout(
+                    Duration::from_secs(options.read_timeout),
+                    cmd.query_async(&mut *conn),
+                )
+                .await
+                .map_err(RedisError::AsyncTimeout)?
+                .map_err(RedisError::Redis)
+            }
+            Self::Single(pool, options) => {
+                let mut conn = pool.get().await.map_err(RedisError::AsyncPool)?;
+                timeout(
+                    Duration::from_secs(options.read_timeout),
+                    cmd.query_async(&mut *conn),
+                )
+                .await
+                .map_err(RedisError::AsyncTimeout)?
+                .map_err(RedisError::Redis)
+            }
+        }
+    }
+
+    /// Forwards the [`Script`] to the pool and loads it asynchronously.
+    /// Returns the SHA1 hash of it.
+    pub async fn load_async(&self, script: &Script) -> Result<String, RedisError> {
+        match self {
+            Self::Cluster(pool, _options) => {
+                let mut conn = pool.get().await.map_err(RedisError::AsyncPool)?;
+                script
+                    .prepare_invoke()
+                    .load_async(&mut *conn)
+                    .await
+                    .map_err(RedisError::Redis)
+            }
+            Self::Single(pool, _options) => {
+                let mut conn = pool.get().await.map_err(RedisError::AsyncPool)?;
+                script
+                    .prepare_invoke()
+                    .load_async(&mut *conn)
+                    .await
+                    .map_err(RedisError::Redis)
+            }
+        }
+    }
```

Comment:
This is only necessary for the multiplexed connection stuff, we don't support it for Async, might as well kill all of it.

This should also make your `create_redis_pools` fn sync again making a bunch of stuff easier.

Reply:
Nice, I'm happy to delete it